### DISCUSSION TITLE
What's the proper way of setting font colors?

### DISCUSSION THREAD

**Discussion by dgrcode** [chakra-ui/chakra-ui#5093] [Q&A] [answered]
I find myself doing a lot of the following:

```js
const primaryFontColor = useColorModeValue("gray.700", "gray.200");
const secondaryFontColor = useColorModeValue("gray.600", "gray.400");
return (
  <>
    <Text color={primaryFontColor}>Some primary text</Text>
    <Text color={secondaryFontColor}>Some secondary text</Text>
    <chakra.li
      key="/my-profile"
      color={secondaryFontColor}
      _hover={{ color: isLightMode ? "var(--chakra-colors-gray-700)" : "var(--chakra-colors-gray-200)" }}
    >
      some link
    </chakra.li>
  </>
)
```

I wish there was a simpler way, maybe just adjusting something on the theme, or using some `colorScheme` somewhere. Maybe there's a way and I don't know about it.

I've looked around the docs and searched on google, but haven't found any solution.

Thanks!

**Comment by keithburgie-rts** [comment, marked as answer]
There's a lot you can do and there are different ways to do it. Read through [the theming section](https://chakra-ui.com/docs/theming/customize-theme) of the docs if you haven't. Any extended theme will be able to access anything from the base Chakra theme (even if you don't import it like I have below).

```javascript
// 1. Import `extendTheme`
import { extendTheme, mode, theme as baseTheme } from "@chakra-ui/react"

// 2. Objects can be created inside the extendTheme function or elsewhere and imported
const colors = {
  primaryFontColor: {
    lightMode: baseTheme.gray["700"],
    darkMode: baseTheme.gray["200"],
  },
  secondaryFontColor: {
    lightMode: baseTheme.gray["600"],
    darkMode: baseTheme.gray["400"],
  },
  plainOldBlue: "blue"
}

// 3. Call `extendTheme` and pass your custom values
const customTheme = extendTheme({
  colors: colors,

  components: {
    Text: {
      baseStyle: props => ({
        color: mode(
          colors.primaryFontColor.lightMode,
          colors.primaryFontColor.darkMode
        )(props),
      }),
      variants: {
        // used as <Text variant="secondary">
        secondary: props => ({
          color: mode(
            colors.primaryFontColor.lightMode,
            colors.primaryFontColor.darkMode
          )(props),
        }),
      },
    },
  },

  styles: {
    global: props => ({
      // Optionally set global CSS styles
      body: {
        color: mode(
          colors.primaryFontColor.lightMode,
          colors.primaryFontColor.darkMode
        )(props),
      },
    }),
  },
})

// 4. Pass the new theme to `ChakraProvider`
<ChakraProvider theme={customTheme}>
  <App />
</ChakraProvider>

// 5. Now you can use these colors in your components
function Usage() {
  return <Box backgroundColor="plainOldBlue">Welcome</Box>
}
```

**Reply by dgrcode** [reply]
Nice! that looks like what I wanted to do. I was hoping I wouldn't need to extend components themes, but that works :)

Thanks!

**Reply by keithburgie-rts** [reply]
@dgrcode Well, if it helps, just keep in mind that you can do a lot or you can do very little with your extended theme. If all you do is create 
```javascript
const customTheme = extendTheme({ 
    colors: { 
        primaryFontColor: "black", 
        secondaryFontColor: "gray"
    } 
})
```

Then you can use `<Text color="primaryFontColor">`

If you really don't want to take that route, you can make reusable components:
```javascript
// In their own files like 'my-components/PrimaryText.jsx'
// Or all together in a file like "my-components/index.js"
export const PrimaryText = (props) => {
  const primaryFontColor = useColorModeValue("gray.700", "gray.200");
  return <Text color={primaryFontColor} {...props} />
}

export const SecondaryText = (props) => {
  const secondaryFontColor = useColorModeValue("gray.600", "gray.400");
  return  <Text color={secondaryFontColor} {...props} />
}

export const StyledLink = (props) => {
  const linkColor = useColorModeValue("blue.600", "red.400");
  const linkHoverColor = isLightMode ? "blue.700" : "red.200"
  return <Link color={linkColor} _hover={{ color: linkHoverColor }} {...props}
}

// Probably in another file...
import { PrimaryText, SecondaryText, StyledLink } from "my-components"

const YourComponent = () => {
  return (
    <>
      <PrimaryText>Some primary text</PrimaryText>
      <SecondaryText>Some secondary text</SecondaryText>
      <StyledLink to="/my-profile">
        some link
      </StyledLink>
    </>
  )
}
```

**Reply by dgrcode** [reply]
I think I like this approach more. I mean having a simpler theme configuration and creating separate components.

Thanks for your help! :)

**Reply by voyagebagage** [reply]
let say that the component one is less team friendly

**Reply by dgrcode** [reply]
@voyagebagage I agree. After a while using the theming and the components approaches for different projects, I think the theming is easier to work with.

I also wanted to share an alternative way, since I just found something interesting looking at the [source code](https://github.com/chakra-ui/chakra-ui/tree/main/packages/theme/src). There's a [`semanticTokens` prop](https://github.com/chakra-ui/chakra-ui/blob/main/packages/theme/src/semantic-tokens.ts) that you can extend in your custom theme like this:

```js
export const fontColors = {
  primary: 'blackAlpha.800',
  secondary: 'blackAlpha.500',
  placeholder: 'blackAlpha.300',
}

const theme = extendTheme({
  semanticTokens: {
    colors: {
      'chakra-body-text': { _light: fontColors.primary },
      'chakra-placeholder-color': { _light: fontColors.placeholder },
    },
  },
})
```
I'm just extending the light version, but it would be simple enough to use the `mode` function as @keithburgie-rts described in the response marked as the answer. Also not adding the `secondary` variant on the `Text` component, which can be seen in the same response.

Extending the `semanticTokens.color` prop will make all the primary color text work just out of the box.

**Comment by ivan-kleshnin** [comment]
❗ The problem with the suggested approach is that is modifies one fundamental property of `Text`, `Heading` etc. text-related components. Meaningly – **color inheritance**.

> `Text` inherits color by default which makes it composable!
> E.g. `Text` inside `Link` will be blue instead of black (assuming white bg) and so on.
> In general, inherited color is a better default than color constants.

To not lose inherticance but also to not pass color values into each `Text, Heading, etc` you rather want them to react to background color change automatically.

And, luckily, it is what Chakra elements do by default with one exception – text components.

https://chakra-ui.com/docs/styled-system/color-mode
`DarkMode` and `LightMode` set right context values but it doesn't set the font color. 

It's easy to built your own component on top of it.

```tsx
type ColorModeProps = {
  mode: "light" | "dark"
  background?: string
  children: React.ReactNode
}

function ColorMode({mode, background, children}: ColorModeProps): JSX.Element {
  if (mode == "light") {
    return <LightMode>
      <Box data-theme="light" color="var(--chakra-colors-chakra-body-text)" background={background}>
        {children}
      </Box>
    </LightMode>
  } else {
    return <LightMode>
      <Box data-theme="dark" color="var(--chakra-colors-chakra-body-text)" background={background}>
        {children}
      </Box>
    </LightMode>
  }
}
```

https://user-images.githubusercontent.com/2128182/180204122-e75321ed-d2d6-41d1-9e34-20caf1b3e63d.mp4

---

Full video-description of the idea: https://youtu.be/hBNuXnDUnLY

**Comment by laduke** [comment]
Good thread.
I'm trying something like this. I haven't cleaned it up yet or picked good color values. Not sure what to name it. Anyways it picks a bg and color based on the colorScheme. No other styling like spacing. You can add that stuff in child elements.

theme
```js
export const Panel1 = {
  baseStyle: {},
  variants: {
    primary: (props) => ({
      bg: mode(`${props.colorScheme}.100`, `${props.colorScheme}.900`)(props),
      color: mode('black.700', 'white.300')(props)
    }),
    secondary: (props) => ({
      bg: mode(`${props.colorScheme}.200`, `${props.colorScheme}.800`)(props),
      color: mode('black.800', 'white.200')(props)
    }),
    tertiary: (props) => ({
      bg: mode(`${props.colorScheme}.300`, `${props.colorScheme}.700`)(props),
      color: mode('black.900', 'white.100')(props)
    })
  },
  defaultProps: {
    variant: 'primary'
  }
}
```

component
```js
export function Panel1 (props) {
  const { colorScheme, variant, children, ...rest } = props

  const styles = useMultiStyleConfig('Panel1', { colorScheme, variant })
  const { colorScheme: y, ...rest2 } = styles
  const x = { ...y, ...rest2 }

  return (
    <Box __css={x} {...rest}>
      {children}
    </Box>
  )
}

const colors = ['whiteAlpha', 'blackAlpha', 'gray', 'red', 'orange', 'green', 'teal', 'blue', 'cyan', 'purple']

Panel1.propTypes = {
  colorScheme: PropTypes.oneOf(colors),
  variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary'])
}
```

Is there a way to get the list of colors out of chakra somehow for my propTypes?

Expecting someone to tell me this is all built in already now. Or I guess I should just add the theming to Box.

**Reply by keithburgie-rts** [reply]
What do you mean by "get the list of colors"? Where would you like to see them?

**Reply by laduke** [reply]
It's off topic,sorry, but I want this array without having to type it out myself: 
`const colors = ['whiteAlpha', 'blackAlpha', 'gray', 'red', 'orange', 'green', 'teal', 'blue', 'cyan', 'purple']`

`Object.keys(extendTheme({}).colors)` works, but there's something wrong with storybook and dynamic propTypes, so oh well. 
The motivation is the react storybook ui which populates from propType:
<img width="248" alt="Screen Shot 2022-07-22 at 10 29 40 AM" src="https://user-images.githubusercontent.com/11598/180493120-c28366c7-d1d8-4c24-9bad-596a8a3916ae.png">


anyways!
